room/integration-tests/testapp/src/androidTest/java/androidx/room/integration/testapp/test/InvalidationTrackerBehavioralTest.java

```diff
@@ -0,0 +1,254 @@
+                        db.counter1Dao().insert(new Counter1());
+
+                        // Use variable delay to detect different kinds of timing-related problems
+                        Thread.sleep(delayMillis, delayNanos);
+
+                        db.runInTransaction(() -> {
+                            db.counter2Dao().insert(new Counter2());
+
+                            // Flag that we have inserted a new value, expect invalidation callback;
+                            // do this as late as possible prior to the end of the transaction;
+                            // this might cause an occasional false negative due to a race,
+                            // where a buggy InvalidationTracker could log successful tracking
+                            // even though the transaction is not completed yet, but it does not
+                            // matter much, as this is an intentionally flaky test; on another run
+                            // it should become apparent that InvalidationTracker is buggy.
+                            latch = new CountDownLatch(1);
+                        });
+
+                        // Use sufficient delay to give invalidation tracker ample time to catch up;
+                        // this would need to be increased if the test has false positives.
+                        if (!latch.await(10L, TimeUnit.SECONDS)) {
+                            // The tracker still has not been called, log an error
+                            missedInvalidations.incrementAndGet();
+                        }
+
+                        latch = null;
+                    }
+                } catch (InterruptedException e) {
+                    throw new RuntimeException(e);
+                }
+            }
+        }).get();
+
+        assertThat("Missed invalidations", missedInvalidations.get(), is(0));
+        assertThat("Spurious invalidations", spuriousInvalidations.get(), is(0));
+    }
+
+    @Database(entities = { Counter1.class, Counter2.class }, version = 1, exportSchema = false)
+    abstract static class DB extends RoomDatabase {
+        static final String NAME = "invalidationtrackerconcurrencytest";
+
+        abstract Counter1Dao counter1Dao();
+        abstract Counter2Dao counter2Dao();
+    }
+
+    @Entity(tableName = Counter1.TABLE_NAME)
+    static final class Counter1 {
+        static final String TABLE_NAME = "counter1";
+
+        @PrimaryKey(autoGenerate = true)
+        long value;
+    }
+
+    @Entity(tableName = Counter2.TABLE_NAME)
+    static final class Counter2 {
+        static final String TABLE_NAME = "counter2";
+
+        @PrimaryKey(autoGenerate = true)
+        long value;
+    }
+
+    @Dao
+    abstract static class Counter1Dao {
+        @Insert
+        abstract void insert(Counter1 entity);
+    }
+
+    @Dao
+    abstract static class Counter2Dao {
+        @Insert
+        abstract void insert(Counter2 entity);
+    }
```

Comment:
Lets just have one `@Dao` call it `CounterDao` with both insert methods to reduce some noise in the test, organization is not as important for test code.